src/storage/v2/durability/snapshot.cpp

```diff
@@ -157,17 +187,378 @@ SnapshotInfo ReadSnapshotInfo(const std::filesystem::path &path) {
+            edge_ref = EdgeRef(&*edge);
+          }
+        }
+        vertex.in_edges.emplace_back(get_edge_type_from_id(*edge_type), &*from_vertex, edge_ref);
+      }
+    }
+
+    // Recover out edges.
+    {
+      spdlog::trace("Recovering outbound edges for vertex {}.", vertex.gid.AsUint());
+      auto out_size = snapshot.ReadUint();
+      if (!out_size) throw RecoveryFailure("Invalid snapshot data!");
+      vertex.out_edges.reserve(*out_size);
+      for (uint64_t j = 0; j < *out_size; ++j) {
+        auto edge_gid = snapshot.ReadUint();
+        if (!edge_gid) throw RecoveryFailure("Invalid snapshot data!");
+
+        auto to_gid = snapshot.ReadUint();
+        if (!to_gid) throw RecoveryFailure("Invalid snapshot data!");
+        auto edge_type = snapshot.ReadUint();
+        if (!edge_type) throw RecoveryFailure("Invalid snapshot data!");
+
+        auto to_vertex = vertex_acc.find(Gid::FromUint(*to_gid));
+        if (to_vertex == vertex_acc.end()) throw RecoveryFailure("Invalid to vertex!");
+
+        EdgeRef edge_ref(Gid::FromUint(*edge_gid));
+        if (items.properties_on_edges) {
+          // If the snapshot was created with a config where properties are not allowed on edges, then the edges has to
+          // be created here
+          if (snapshot_has_edges) {
+            auto edge = edge_acc.find(Gid::FromUint(*edge_gid));
+            if (edge == edge_acc.end()) throw RecoveryFailure("Invalid edge!");
+            edge_ref = EdgeRef(&*edge);
+          } else {
+            auto [edge, inserted] = edge_acc.insert(Edge{Gid::FromUint(*edge_gid), nullptr});
+            edge_ref = EdgeRef(&*edge);
+          }
+        }
+        vertex.out_edges.emplace_back(get_edge_type_from_id(*edge_type), &*to_vertex, edge_ref);
+        // Increment edge count. We only increment the count here because the
+        // information is duplicated in in_edges.
+        edge_count++;
+      }
+    }
+    ++vertex_it;
+  }
+  spdlog::info("Partial connectivities are recovered.");
+  return {edge_count, highest_edge_gid, first_vertex_gid};
+}
+
+template <typename TFunc>
+void RecoverOnMultipleThreads(const size_t thread_count, const TFunc &func, const std::vector<BatchInfo> &batches) {
+  utils::Synchronized<std::optional<RecoveryFailure>, utils::SpinLock> maybe_error{};
+  {
+    std::atomic<uint64_t> batch_counter = 0;
+    std::vector<std::jthread> threads;
+    threads.reserve(thread_count);
+
+    for (auto i{0U}; i < thread_count; ++i) {
+      threads.emplace_back([&func, &batches, &maybe_error, &batch_counter]() {
+        while (!maybe_error.Lock()->has_value()) {
+          const auto batch_index = batch_counter++;
+          if (batch_index >= batches.size()) {
+            return;
+          }
+          const auto &batch = batches[batch_index];
+          try {
+            func(batch_index, batch);
+          } catch (RecoveryFailure &failure) {
+            *maybe_error.Lock() = std::move(failure);
```

Comment:
What happens when error occurs? So, if I got this correctly, it could happen that there is error in one batch of snapshot, but it will keep recovering data until it is finished in other batches until all are finished, right?

Reply:
Yes. I did that intentionally, because I wanted to keep the happy path optimal: we strongly expect the snapshot loading work without issues, therefore checking whether the processing of a batch should stop is not necessary. Checking it in every iteration of the inner loop (which deserializes the objects) would result in a lot of synchronization between threads.

The current solution gives us the:
1. better performance on the happy path
2. logically correct solution in any case
3. the user might wait a bit longer to actually receive the error, but that is not the usual use-case in my opinion.

Reply:
Okay, I am okay with that we don't interrupt batches that are working.

Does it make sense to check for other batches that are about to start if there is an error and not start them?
```
 try {
  if (*maybe_error.Lock()) return;
  func(batch_index, batch);
} catch (RecoveryFailure &failure) {
  *maybe_error.Lock() = std::move(failure);
```

Reply:
We can do, but the code between the beginning of `while` and the `try` block is expected to be fast and the `while` loop has the same check as a loop condition. In general I don't think it has any practical benefit to check this twice right after each other.

Reply:
You are right, I think we can skip this check.